### PR TITLE
Implement `assert.match()` and `assert.doesNotMatch()`

### PR DESCRIPTION
**Pull request by RaisinTen**
Signed-off-by: Darshan Sen <darshan.sen@postman.com>

### CODE REVIEW

**Comment by RaisinTen** [comment]
cc @BridgeAR

File: assert.js [hunk 1]
```diff
@@ -596,6 +596,51 @@ assert.ifError = function ifError(err) {
...
 
+// Currently in sync with Node.js lib/assert.js
+// https://github.com/nodejs/node/commit/2a871df3dfb8ea663ef5e1f8f62701ec51384ecb
+function internalMatch(string, regexp, message, fn) {
+  if (!isRegExp(regexp)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'regexp', 'RegExp', regexp
+    );
+  }
+  const match = fn.name === 'match';
```
Comment (ljharb) [inline assert.js:607]: function names are configurable, and some browsers don't support function names at all (IE 9-11). Instead, this should probably just be a hardcoded string.

File: assert.js [hunk 2]
```diff
@@ -596,6 +596,51 @@ assert.ifError = function ifError(err) {
...
+  if (typeof string !== 'string' ||
+      regexp.test(string) !== match) {
```
Comment (ljharb) [inline assert.js:609]: this is vulnerable to `delete RegExp.prototype.test` (which node core is not); you'll want to call-bind that function at module level, like this:
```js
const test = require('call-bind/callBound')('RegExp.prototype.test');
```
and then you can do `test(regexp, string)` here.

File: assert.js (hunk 1 above)
Comment (RaisinTen) [reply assert.js]: True, updated accordingly. However, I'm wondering if we should update the rest of the `fn.name`s in the file too, like in https://github.com/browserify/commonjs-assert/blob/bba838e9ba9e28edf3127ce6974624208502f6bc/assert.js#L337 in a follow-up PR?

File: assert.js (hunk 2 above)
Comment (RaisinTen) [reply assert.js]: @ljharb Thanks, done!

File: assert.js (hunk 1 above)
Comment (ljharb) [reply assert.js]: Yes, we definitely should. Relying on function names is not robust.

File: assert.js [hunk 3]
```diff
@@ -596,6 +596,52 @@ assert.ifError = function ifError(err) {
...
+// Currently in sync with Node.js lib/assert.js
+// https://github.com/nodejs/node/commit/2a871df3dfb8ea663ef5e1f8f62701ec51384ecb
+function internalMatch(string, regexp, message, fn, fnName) {
+  if (!isRegExp(regexp)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'regexp', 'RegExp', regexp
+    );
+  }
+  const match = fnName === 'match';
+  const test = require('call-bind/callBound')('RegExp.prototype.test');
```
Comment (ljharb) [inline assert.js:608]: i meant that as quasi pseudocode :-) the require and this line should both be done at module level (`var callBound = require('call-bind/callBound');`) not inside this runtime function.

Comment (RaisinTen) [reply assert.js]: @ljharb :facepalm: of course! I moved it to the module level and named it `RegExpPrototypeTest` instead (like we name primordials in Node.js) and used it wherever the `test` functionality was being used. Not very sure why we should use `var` though, given that it is very sparingly used in this module. Could u please explain?

**Comment by RaisinTen** [comment]
Btw, I'd like to mention that when I ran `npm run build && npm run test` to test this locally, it passes all the tests but when I run `npm run build && npm run test:browsers:local` to test this on Safari, only the `test-assert-match.js` test fails for some reason and I can't see a stack trace to debug this. What's the general debugging workflow for this?

File: test/parallel/test-assert-match.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+require('../common');
+const assert = require('../../assert');
+
+// Multiple assert.match() tests.
+{
+  assert.throws(
+    () => assert.match(/abc/, 'string'),
+    {
+      code: 'ERR_INVALID_ARG_TYPE',
+      message: 'The "regexp" argument must be of type RegExp. Received type string'
```
Comment (RaisinTen) [inline test/parallel/test-assert-match.js:15]: There is a slight difference in the message generated by this implementation and the one in Node.js. It should be actually be:
`The "regexp" argument must be an instance of RegExp. Received type string ('string')`
I think this should be updated in a separate PR as it touches the `ERR_INVALID_ARG_TYPE` error type which is not super-related to `assert.match()`

**Review by BridgeAR — APPROVED** [review]

File: assert.js (hunk 3 above)
Comment (ljharb) [reply assert.js]: This entire package is a browser shim, for older engines; if other things are using const than obv you can use it too, but probably none of it should be.

File: assert.js [hunk 5]
```diff
@@ -37,6 +37,8 @@ const { isPromise, isRegExp } = require('util/').types;
 const objectAssign = Object.assign ? Object.assign : require('es6-object-assign').assign;
 const objectIs = Object.is ? Object.is : require('object-is');
 
+const RegExpPrototypeTest = require('call-bind/callBound')('RegExp.prototype.test');
```
Comment (ljharb) [inline assert.js:40]: call-bind needs to be added to package.json also

File: assert.js (hunk 3 above)
Comment (RaisinTen) [reply assert.js]: Cool, I'll do that in a follow-up PR :+1:

File: assert.js (hunk 5 above)
Comment (RaisinTen) [reply assert.js:40]: @ljharb Thanks, added

**Review by ljharb — APPROVED** [review]

**Comment by RaisinTen** [comment]
@ljharb @BridgeAR Is there a minimum waiting time before we can land this PR and make a new release? I couldn't find any documentation for the process.

**Comment by RaisinTen** [comment]
ping @goto-bus-stop ^

**Comment by ljharb** [comment]
@RaisinTen pinging them multiple times is just annoying, it's not going to make anything happen faster. They'll get to it when they have time, and not before.

**Review by AlxKrispy — APPROVED** [review]
